python/chemiscope/explore.py

```diff
@@ -98,13 +109,33 @@ def soap_kpca_featurize(frames):
     if properties is None:
         properties = {}
 
+    # Pick frames and properties related to the environments if provided
+    if environments is not None:
+        # Sort environments by structure id and atom id
+        environments = sorted(environments, key=lambda x: (x[0], x[1]))
+
+        # Check structure indexes
+        unique_structures = list({env[0] for env in environments})
+        if any(index >= len(frames) for index in unique_structures):
+            raise IndexError(
+                "Some or more indices are greater than the length of the frames"
+            )
+
+        if len(unique_structures) != len(frames):
+            # Pick frames corresponding to the environments
+            frames = [frames[index] for index in unique_structures]
```

Comment:
I'm not sure this is correct: after running it, the indexes in environments will no longer match the positions of items in the `frames` list.

Reply:
If this is an issue with DScribe, it should be handled in the DScribe adapter.

Reply:
Done